# Assignment: 


## 1. Weather model

* First, load the dataset from the **weatherinszeged** table from Thinkful's database.
* Like in the previous checkpoint, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why? 

In [35]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm
from scipy import stats
from sklearn import linear_model
%matplotlib inline
sns.set()

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

In [36]:
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
weather = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

In [37]:
# Target variable: difference between apparenttemperature and temperature
Y = weather['apparenttemperature'] - weather['temperature']

# Feature set
X = weather[['humidity','windspeed']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Fri, 31 May 2019   Prob (F-statistic):               0.00
Time:                        15:13:48   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- The R-squared and Adjusted R-squared values is 0.288. 
- 0.288 is a relatively low R-squared value, as only 28.8% of the variance in the target variable is explained by the model, leaving 71.2% of the variation in the target variable left unexplained. 
    - In some cases, the target variable is difficult to explain and 0.288 could be a relatively good score. 

Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?

In [38]:
# Engineering a new feature for the interaction between humidity and windspeed 
weather['humidity_windspeed_int'] = weather.humidity * weather.windspeed

# Target variable: difference between apparenttemperature and temperature
Y = weather['apparenttemperature'] - weather['temperature']

# Feature set
X = weather[['humidity','windspeed', 'humidity_windspeed_int']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Fri, 31 May 2019   Prob (F-statistic):               0.00
Time:                        15:13:48   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                      0.0839      0.033      2.511      0.012       0.018       0.149
humidity                   0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                  0.0905      0.002     36.797      0.000       0.086       0.095
humidity_windspeed_int    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* Both R-squared and adjusted R-squared values are 0.341. Including the interaction term increased the adjusted R-squared value by 0.053 points. This is an improvement upon the previous model.

Add visibility as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [39]:
# Target variable: difference between apparenttemperature and temperature
Y = weather['apparenttemperature'] - weather['temperature']

# Feature set
X = weather[['humidity','windspeed', 'visibility']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Fri, 31 May 2019   Prob (F-statistic):               0.00
Time:                        15:13:48   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- R-squared of the model is 0.304 and the adjusted R-squared is 0.303. Both R-squared and adjusted R-squared increased compared to the first model. Adding visibility to the model improved the adjusted R-squared by 0.015 point. However, this improvement is lower than the one that is achieved by putting the interaction term in the model. Hence, interaction term contributes more to the model. From this point of view, interaction term can be regarded as more important than the visibility in explaining the information in the target.

Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor

- As we know, the lower values the better for information criterions like AIC and BIC. The first model's AIC and BIC values are 340900. The second model's are 333400 and the third model's are 338800. Among these, the model with the lowest AIC and BIC scores is the second model. So, the best of these three models is the second one.

## Exercise 2: House Prices Model

In [40]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm
from scipy import stats
from sklearn import linear_model
%matplotlib inline
sns.set()

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

In [41]:
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
house_prices_df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC

In [42]:
# first, organizing the numerical and non-numerical data:
# Numerical features:
numerical = house_prices_df.dtypes[house_prices_df.dtypes != "object"].index
numerical.ravel()

# Non-numerical features:
non_numerical = house_prices_df.dtypes[house_prices_df.dtypes == "object"].index
non_numerical.ravel()

array(['mszoning', 'street', 'alley', 'lotshape', 'landcontour',
       'utilities', 'lotconfig', 'landslope', 'neighborhood',
       'condition1', 'condition2', 'bldgtype', 'housestyle', 'roofstyle',
       'roofmatl', 'exterior1st', 'exterior2nd', 'masvnrtype',
       'exterqual', 'extercond', 'foundation', 'bsmtqual', 'bsmtcond',
       'bsmtexposure', 'bsmtfintype1', 'bsmtfintype2', 'heating',
       'heatingqc', 'centralair', 'electrical', 'kitchenqual',
       'functional', 'fireplacequ', 'garagetype', 'garagefinish',
       'garagequal', 'garagecond', 'paveddrive', 'poolqc', 'fence',
       'miscfeature', 'saletype', 'salecondition'], dtype=object)

In [43]:
# Make a copy of the orginal df to work with
house_prices_df2 = house_prices_df.copy()

# converting all non-numerical columns to numerical, by one-hot encoding them
# this will make the whole house_prices_df2 have continuous features
for variable in non_numerical:
    house_prices_df2 = pd.concat([house_prices_df, pd.get_dummies(house_prices_df2[variable], prefix=variable, drop_first=True)], axis=1)
    

In [44]:
# Finding the top 10 most correlated features:

np.abs(house_prices_df2.corr().loc[:,'saleprice']).sort_values(ascending=False).head(10).index


Index(['saleprice', 'overallqual', 'grlivarea', 'garagecars', 'garagearea',
       'totalbsmtsf', 'firstflrsf', 'fullbath', 'totrmsabvgrd', 'yearbuilt'],
      dtype='object')

In [45]:
# The feature set:
X = house_prices_df2[['overallqual', 'grlivarea', 'garagecars', 'garagearea',
       'totalbsmtsf', 'firstflrsf', 'fullbath', 'totrmsabvgrd', 'yearbuilt']]

# target is saleprice
Y = house_prices_df2.saleprice

import statsmodels.api as sm

# We need to manually add a constant
# in statsmodels' sm
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.770
Model:                            OLS   Adj. R-squared:                  0.769
Method:                 Least Squares   F-statistic:                     540.3
Date:                Fri, 31 May 2019   Prob (F-statistic):               0.00
Time:                        15:13:49   Log-Likelihood:                -17470.
No. Observations:                1460   AIC:                         3.496e+04
Df Residuals:                    1450   BIC:                         3.501e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -7.585e+05   9.15e+04     -8.294      0.000   -9.38e+05   -5.79e+05
overallqual   2.092e+04   1164.267     17.971      0.000    1.86e+04    2.32e+04
grlivarea       51.3679      4.263     12.051      0.000      43.006      59.730
garagecars    1.063e+04   3065.576      3.469      0.001    4620.797    1.66e+04
garagearea      14.5088     10.384      1.397      0.163      -5.861      34.879
totalbsmtsf     19.2466      4.324      4.451      0.000      10.765      27.728
firstflrsf      13.9481      4.965      2.810      0.005       4.210      23.687
fullbath     -5410.3114   2684.921     -2.015      0.044   -1.07e+04    -143.567
totrmsabvgrd   -86.6391   1126.726     -0.077      0.939   -2296.827    2123.549
yearbuilt      345.3629     47.886      7.212      0.000     251.430     439.295
==============================================================================
Omnibus:                      470.067   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            52956.991
Skew:                          -0.413   Prob(JB):                         0.00
Kurtosis:                      32.493   Cond. No.                     2.77e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.77e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- The R-squared value is 0.770 and the adjusted R-squared value is 0.769
- The F-statistic is 540.3 and the associated p-value is 0.00. 
- AIC and BIC of the model is 34960 and 35010, respectively

Do you think your model is satisfactory? If so, why?

- I think the model does an adequate job, as 77% of the variance in the target variable is explained. However, 23% of the variance is not explained, and there is room for improvement. 

In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.

In [46]:
# Engineering new features: 
house_prices_df2['totalsf'] = house_prices_df2['totalbsmtsf'] + house_prices_df2['firstflrsf'] + house_prices_df2['secondflrsf']

house_prices_df2['int_over_sf'] = house_prices_df2['totalsf'] * house_prices_df2['overallqual']

# The feature set:
X = house_prices_df2[['overallqual', 'grlivarea', 'garagecars', 'garagearea',
       'totalsf', 'fullbath', 'totrmsabvgrd', 'yearbuilt', 'int_over_sf']]

# target is saleprice
Y = house_prices_df2.saleprice

import statsmodels.api as sm

# We need to manually add a constant
# in statsmodels' sm
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()



<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.780
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     571.7
Date:                Fri, 31 May 2019   Prob (F-statistic):               0.00
Time:                        15:50:16   Log-Likelihood:                -17438.
No. Observations:                1460   AIC:                         3.490e+04
Df Residuals:                    1450   BIC:                         3.495e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -6.233e+05      9e+04     -6.926      0.000      -8e+05   -4.47e+05
overallqual   7401.4768   1934.701      3.826      0.000    3606.364    1.12e+04
grlivarea       21.2487      5.443      3.904      0.000      10.572      31.925
garagecars    1.376e+04   3021.179      4.556      0.000    7837.047    1.97e+04
garagearea       6.2421     10.204      0.612      0.541     -13.774      26.258
totalsf         -9.0002      5.265     -1.709      0.088     -19.328       1.327
fullbath     -2689.9642   2638.399     -1.020      0.308   -7865.451    2485.522
totrmsabvgrd  1383.3086   1115.653      1.240      0.215    -805.159    3571.776
yearbuilt      320.5223     46.817      6.846      0.000     228.687     412.358
int_over_sf      5.2437      0.626      8.376      0.000       4.016       6.472
==============================================================================
Omnibus:                     1176.375   Durbin-Watson:                   1.989
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           242553.075
Skew:                          -2.855   Prob(JB):                         0.00
Kurtosis:                      65.885   Cond. No.                     1.74e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.74e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?


- R-squared and adjusted R-squared of the model are 0.780 and 0.779 respectively. These are improvements upon the first model.
- F statistic and its associated p-value are 571.7 and 0.00 respectively. This also indicates that the second model is better than the first one.
- AIC and BIC of the model are 34900 and 34950 respectively. These values are lower than the ones of the first model. Hence, the second model is better than the first model.
- Considering all of the metrics above, our second model has a better goodness of fit than the first model.